src/Features/Core/Portable/SolutionCrawler/IDocumentTrackingServiceExtensions.cs

```diff
@@ -30,5 +33,43 @@ public static ImmutableArray<Document> GetVisibleDocuments(this IDocumentTrackin
                       .WhereNotNull()
                       .Distinct()
                       .ToImmutableArray();
+
+        public static Task DelayIfNonVisibleAsync(this IDocumentTrackingService service, Document document, TimeSpan timeSpan, CancellationToken cancellationToken)
+            => DelayIfNonVisibleAsync(service, ImmutableArray.Create(document), timeSpan, cancellationToken);
+
+        public static async Task DelayIfNonVisibleAsync(this IDocumentTrackingService service, ImmutableArray<Document> documents, TimeSpan timeSpan, CancellationToken cancellationToken)
+        {
+            // Only add a delay if we have access to a service that will tell us when the documents become visible or not.
+            if (!service.SupportsDocumentTracking)
+                return;
+
+            // if any of the docs are visible, then no delay.
+            var visibleDocuments = service.GetVisibleDocuments();
+            var anyVisible = documents.Select(d => d.Id).Intersect(visibleDocuments).Any();
+            if (anyVisible)
+                return;
+
+            // Listen to when the active document changed so that we startup work on a document once it becomes visible.
+
+            var delayTask = Task.Delay(timeSpan, cancellationToken);
+            var visibilityChangedTaskSource = new TaskCompletionSource<bool>();
+            service.ActiveDocumentChanged += OnActiveDocumentChanged;
+
+            try
+            {
+                await Task.WhenAny(delayTask, visibilityChangedTaskSource.Task).ConfigureAwait(false);
+            }
+            finally
+            {
+                service.ActiveDocumentChanged -= OnActiveDocumentChanged;
+            }
+
+            return;
+
+            void OnActiveDocumentChanged(object? sender, DocumentId? e)
+            {
+                visibilityChangedTaskSource.SetResult(true);
```

Comment:
Should this check `service.VisibleDocuments()` to ensure the change is actually important?

Reply:
i'll doc it, but i'm worried about this service not being entirely reliable.  since switching active doc is rare, i'm fine with getting hte benefit as long as teh user is editing in one doc, and then when you switch we do a bit of extra work.